lib/charms/opensearch/v0/opensearch_performance_profile.py

```diff
@@ -0,0 +1,178 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Represents the performance profile of the OpenSearch cluster.
+
+The main goals of this library is to provide a way to manage the performance
+profile of the OpenSearch cluster.
+
+There are two ways the charm can learn about its profile and when it changes:
+1) If this is the MAIN_ORCHESTRATOR: config-changed -> the user has switched the profile directly
+2) If not the MAIN_ORCHESTRATOR: peer-cluster-relation-changed -> the main orchestrator has
+                                 switched the profile
+
+The charm will then apply the profile and restart the OpenSearch service if needed.
+"""
+import logging
+
+import ops
+from charms.opensearch.v0.constants_charm import PERFORMANCE_PROFILE
+from charms.opensearch.v0.models import (
+    DeploymentType,
+    OpenSearchPerfProfile,
+    PerformanceType,
+)
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchHttpError
+from charms.opensearch.v0.opensearch_internal_data import Scope
+from ops.framework import EventBase, EventSource
+
+# The unique Charmhub library identifier, never change it
+LIBID = "8b7aa39016e748ea908787df1d7fb089"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+logger = logging.getLogger(__name__)
+
+
+class _ApplyProfileTemplatesOpenSearch(EventBase):
+    """Attempt to apply the profile templates.
+
+    The main reason to have a separate event, is to be able to wait for the cluster to restart.
+    In this case, deferring this event does not defer any major other event.
+    """
+
+
+class OpenSearchPerformance(ops.Object):
+    """Base class for OpenSearch charms."""
+
+    _apply_profile_templates_event = EventSource(_ApplyProfileTemplatesOpenSearch)
+
+    def __init__(self, charm: ops.charm.CharmBase = None):
+        super().__init__(charm, None)
+        self.charm = charm
+        self.peers_data = self.charm.peers_data
+        self.framework.observe(
+            self._apply_profile_templates_event, self._on_apply_profile_templates
+        )
+        self._apply_profile_templates_has_been_called = False
+
+    @property
+    def current(self) -> OpenSearchPerfProfile | None:
+        """Return the current performance profile.
+
+        The profile is saved as a string in the charm peer databag.
+        """
+        if not self.peers_data.get(Scope.UNIT, PERFORMANCE_PROFILE):
+            return None
+        return OpenSearchPerfProfile.from_dict(
+            {"typ": self.peers_data.get(Scope.UNIT, PERFORMANCE_PROFILE)}
+        )
```

Comment:
nit for conciseness:
```suggestion
        if not (profile := self.peers_data.get(Scope.UNIT, PERFORMANCE_PROFILE)):
            return None
        return OpenSearchPerfProfile.from_dict({"typ": profile})
```